### PR TITLE
Heat exchanger sizing changes

### PR DESCRIPTION
**Pull request by drewyh**
Pull request overview
---------------------
This pull request changes the methodology for sizing the demand side flow rate for fluid to fluid heat exchangers. Previously the demand side flow rate was set equal to the supply side flow rate. However this meant that it did not take into account any differences in the fluid or temperature differences of the two loops. This is changed in this pull request to take this into account and includes tests for sizing of the HX.

The changes require that the demand side loop have a Sizing:Plant object if the demand side flow is to be autosized. This is a new requirement so may cause some errors in existing files.

This addresses the suggestion made in #6064.

### Queries
- When auto-sizing the HX, the temperature `DataGlobals::InitConvTemp` is used for calculating specific heat and density, which was based on the existing code for auto-sizing UA. There seems to be some inconsistency on how this is addressed in other plant items, e.g. boiler uses `DataGlobals::CWInitConvTemp` for density, and boiler outlet temperature for specific heat capacity. Is there any guidance on how this should be standardised across plant items?
- I'm currently unsure how to set up a unit test for differing fluids between the supply and demand loops, any pointers would be helpful as this really should be part of the testing.
- The HX model is run to determine the nominal capacity, however, if the HX is autosized for flows and UA then this may not be necessary?

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [ ] Code style (parentheses padding, variable names)
 - [ ] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check -- verify this
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add rules to spreadsheet
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Documentation changes in place
 - [ ] Changed docs build successfully
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by nrel-bot-3** [comment]
@drewyh @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@drewyh @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@drewyh @lgentile it has been 14 days since this pull request was last updated.

**Review by rraustad — APPROVED** [review]
This change looks appropriate. Comments need to be answered.

File: src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc [hunk 1]
```diff
@@ -827,20 +831,47 @@ namespace PlantHeatExchangerFluidToFluid {
...
-			}
-			if ( PlantFinalSizesOkayToReport ) {
-				ReportSizingOutput( "HeatExchanger:FluidToFluid", FluidHX( CompNum ).Name,
-					"Loop Demand Side Design Fluid Flow Rate [m3/s]", FluidHX( CompNum ).DemandSideLoop.DesignVolumeFlowRate );
+			if (PltSizNumDmdSide > 0) {
+				if ( tmpSupSideDesignVolFlowRate > SmallWaterVolFlow ) {
+					// calculate the supply side capacity per volume flow rate
+					tmpDeltaTSupLoop = PlantSizData(PltSizNumSupSide).DeltaT;
+					Cp = GetSpecificHeatGlycol(PlantLoop(FluidHX(CompNum).SupplySideLoop.LoopNum).FluidName, DataGlobals::InitConvTemp, PlantLoop(FluidHX(CompNum).SupplySideLoop.LoopNum).FluidIndex, RoutineName);
+					rho = GetDensityGlycol(PlantLoop(FluidHX(CompNum).SupplySideLoop.LoopNum).FluidName, DataGlobals::InitConvTemp, PlantLoop(FluidHX(CompNum).SupplySideLoop.LoopNum).FluidIndex, RoutineName);
```
Comment (rraustad) [inline src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc]: For HW loops this should use HWInitConvTemp. Is this HX allowed in cold or hot water loops? If so then the loop type would need to be tested to use the correct InitConvTemp initialization.

File: tst/EnergyPlus/unit/PlantHeatExchangerFluidToFluid.unit.cc [hunk 2]
```diff
@@ -2490,4 +2491,84 @@ namespace EnergyPlus {
...
+		// set demand loop fluid to 50% glycol - unsure how this should work
+		DataSizing::PlantSizData(2).DeltaT = 10.0;
+		PlantHeatExchangerFluidToFluid::SizeFluidHeatExchanger(1);
+		EXPECT_EQ(PlantHeatExchangerFluidToFluid::FluidHX(1).SupplySideLoop.DesignVolumeFlowRate, 1.0);
+		EXPECT_NEAR(PlantHeatExchangerFluidToFluid::FluidHX(1).DemandSideLoop.DesignVolumeFlowRate, 1.1283886, 0.000001);
+		*/
+
+		PlantHeatExchangerFluidToFluid::FluidHX.deallocate();
+		DataSizing::PlantSizData.deallocate();
+		DataPlant::PlantLoop.deallocate();
```
Comment (rraustad) [inline tst/EnergyPlus/unit/PlantHeatExchangerFluidToFluid.unit.cc]: We should start requiring that this type of clean up not be allowed, or at least if they are required a followup issue (for those not adept at clear_state creation) to create a clear_state where needed.

**Comment by nrel-bot-3** [comment]
@drewyh @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@drewyh it has been 8 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@drewyh it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@drewyh it has been 8 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@drewyh @lgentile it has been 16 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@drewyh @lgentile it has been 36 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@drewyh @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@drewyh @lgentile it has been 85 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@drewyh it has been 7 days since this pull request was last updated.

**Comment by nrel-bot-2c** [comment]
@drewyh it has been 7 days since this pull request was last updated.

**Comment by Myoldmopar** [comment]
@drewyh first thanks for submitting this (a while back!).  When features get submitted directly as a pull request without discussion ahead of time, it's tough to get it into our routine review processes and workflows, so this has lingered for sure.

I see that this comes from a request from @EnergyArchmage.  I could start there and ask if this meets your requirements on that request, @EnergyArchmage. 

I also see that @rraustad has looked this over and generally liked the changes but had at least question that needs a response.

Let me know if you have time to spend bringing this back up so we can consider getting it into the codebase.  If not, I'll close the PR, but that doesn't mean we can't reopen it at a later time.

Thanks!

**Comment by Myoldmopar** [comment]
Per the above message, closing for now, but happy to reopen.

File: src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc [hunk 3]
```diff
@@ -959,24 +965,84 @@ namespace PlantHeatExchangerFluidToFluid {
...
+
+                    if (supplyLoopType == HeatingLoop) {
+                        supplyLoopInitTemp = HWInitConvTemp;
+                    }
+                    else if (supplyLoopType == CondenserLoop || supplyLoopType == CoolingLoop) {
+                        supplyLoopInitTemp = CWInitConvTemp;
+                    }
+                    else {
+                        assert(false);
+                    }
```
Comment (rraustad) [inline src/EnergyPlus/PlantHeatExchangerFluidToFluid.cc:984]: Correct initialization temp used here, I just didn't notice when this was corrected.

**Comment by rraustad** [comment]
Now my only question is the impact these changes have on the simulation. A before and after comparison of results for a defect file or example file that shows what these changes correct.

**Comment by EnergyArchmage** [comment]
I like these changes.  Seems like it was a thorough contribution, code, docs, unit test.  Unfortunate that this fell thru the cracks.

**Pull request closed without merging**